src/backend/distributed/planner/multi_master_planner.c

```diff
@@ -203,10 +274,45 @@ BuildSelectStatement(Query *masterQuery, List *masterTargetList, CustomScan *rem
 		topLevelPlan = &remoteScan->scan.plan;
 	}
 
-	/* (3) add a sorting plan if needed */
-	if (masterQuery->sortClause)
+	/*
+	 * (3) create distinct plan if needed.
+	 *
+	 * distinct on() requires sort + unique plans. If already has order by
+	 * clause we append distinct clauses to the end of it. Postgresql requires
+	 * that if both distinct on() and order by exists, ordering shall start
+	 * on distinct clauses. Therefore we can safely append distinct clauses to
+	 * the end of order by clauses. Although the same column may appear more
+	 * than once in order by clauses, created plan uses only one instance, for
+	 * example order by a,b,a,a,b,c is translated to equivalent order by a,b,c.
+	 *
+	 * If the query has distinct clause but not distinct on, we first create
+	 * distinct plan that is either HashAggreate or Sort + Unique plans depending
+	 * on hashable property of columns in distinct clause. If there is order by
+	 * clause, it is handled after distinct planning.
+	 */
+	if (masterQuery->hasDistinctOn)
+	{
+		ListCell *distinctCell = NULL;
```

Comment:
Thinking about this a bit more, do we need this? PostgreSQL requires distinct column to be added to order by clause if there is order by. If there is no order by, then the order is not important. Can't we just replace whole block with;

```
	if (masterQuery->distinctClause && !masterQuery->hasDistinctOn)
	{
		Plan *distinctPlan = BuildDistinctPlan(masterQuery, topLevelPlan);
		topLevelPlan = distinctPlan;
	}
```

Reply:
Unique requires ordering, otherwise it won't work. Keeping this as it is per our discussion